typing/ctype.ml

```diff
@@ -4361,46 +4361,52 @@ let nondep_variants = TypeHash.create 17
 let clear_hash ()   =
   TypeHash.clear nondep_hash; TypeHash.clear nondep_variants
 
-let rec nondep_type_rec ?(expand_private=false) env id ty =
+exception Nondep_cannot_erase of Ident.t
+
+let rec nondep_type_rec ?(expand_private=false) env ids ty =
   let expand_abbrev env t =
     if expand_private then expand_abbrev_opt env t else expand_abbrev env t
   in
   match ty.desc with
     Tvar _ | Tunivar _ -> ty
-  | Tlink ty -> nondep_type_rec env id ty
+  | Tlink ty -> nondep_type_rec env ids ty
   | _ -> try TypeHash.find nondep_hash ty
   with Not_found ->
     let ty' = newgenvar () in        (* Stub *)
     TypeHash.add nondep_hash ty ty';
     ty'.desc <-
       begin match ty.desc with
       | Tconstr(p, tl, _abbrev) ->
-          if Path.isfree id p then
-            begin try
-              Tlink (nondep_type_rec ~expand_private env id
-                       (expand_abbrev env (newty2 ty.level ty.desc)))
-              (*
-                 The [Tlink] is important. The expanded type may be a
-                 variable, or may not be completely copied yet
-                 (recursive type), so one cannot just take its
-                 description.
-               *)
-            with Cannot_expand | Unify _ ->
-              raise Not_found
-            end
-          else
-            Tconstr(p, List.map (nondep_type_rec env id) tl, ref Mnil)
-      | Tpackage(p, nl, tl) when Path.isfree id p ->
+          begin match Path.find_free_opt ids p with
+          | Some id ->
+              begin try
+                Tlink (nondep_type_rec ~expand_private env ids
+                         (expand_abbrev env (newty2 ty.level ty.desc)))
+                (*
+                   The [Tlink] is important. The expanded type may be a
+                   variable, or may not be completely copied yet
+                   (recursive type), so one cannot just take its
+                   description.
+                 *)
+              with Cannot_expand | Unify _ ->
+                raise (Nondep_cannot_erase id)
+              end
+          | None ->
+              Tconstr(p, List.map (nondep_type_rec env ids) tl, ref Mnil)
+          end
```

Comment:
Not sure which change you're referring too (the one above or the one below?), but neither of them is here to avoid modifying the type. It's just "plumbing", i.e. the behavior doesn't change from the previous version, it's just that instead of `if Path.is_free` we have `match Path.find_free_opt`

Reply:
Ah, Indeed, sorry!